### DISCUSSION TITLE
Error codes vs assertion failures/segfaults

### DISCUSSION THREAD

**Discussion by zsugabubus** [openssl/openssl#21896] [Q&A]
Across code base, only an error code is returned when fundamental preconditions does not hold. What is the reason behind this?

A few examples:

https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/crypto/evp/evp_enc.c#L693-L697

When I have messed up decryption and encryption routines I would like my program to utterly crash. No? It is not a tiny bit error that I can handle gracefully and adjust parameters hoping next time it will succeed. It is some serious logic failure.

https://github.com/openssl/openssl/blob/2c61a670ebf2f1923a3bd2ef0ee4b2fa6261eaeb/crypto/bio/bio_lib.c#L325-L341

When something is not implemented, it will not be at the next time... and neither at the next time... and the next. All these error codes surely *require* modifications at code level... so why return them? I’m not an expert but if something is not compatible together or not implemented, what is the expected what should I do after checking for `-2`? Using nothing instead or relying on worst/weaker alternatives? Why check for NULLs anyway?

**Comment by nhorman** [comment]
You may want your application to crash, but not every user does (arguably, most users don't).  openssl (and typically almost every library) endeavours to return an error according to a specification so that a using application can take appropriate action, be that retrying, or in your case issuing a fault.  In the case you are looking at:
https://www.openssl.org/docs/man1.1.1/man3/EVP_EncryptUpdate.html
returns 1 for failure and 0 for success, pushing errors to the error stack so that your application can determine what went wrong.

If your goal is to crash your application, thats easy enough to do.  In your code:

```
assert(!EVP_EncryptUpdate(ctx, out, &outl, in, inl));
```

A better solution might be:
```
if (!EVP_EncryptUpdate(ctx, out, &outl, in, &inl)) {
     ERR_error_get_all(&file, &line, &func, &data, &flags);
      fprintf(stderr, "%s: %d, %s, %s, %d, file, line func, data, flags);
      abort();
}
```
Then your application will print a report about what went wrong so you can fix it.

But if we just ignored all error conditions, people would have a nightmare of a time trying to fix anything.
